proto/symge/symge-demo.cpp

```diff
@@ -0,0 +1,387 @@
+    for (unsigned r = 0; r<m.nrow(); ++r) {
+        o << '|';
+        for (unsigned c = 0; c<m.ncol(); ++c) {
+            if (c==m.augcol()) o << " | ";
+            o << std::setw(12) << m[r][c];
+        }
+        o << " |\n";
+    }
+    return o;
+}
+
+template <typename Sep, typename V>
+struct sepval_t {
+    const Sep& sep;
+    const V& v;
+
+    sepval_t(const Sep& sep, const V& v): sep(sep), v(v) {}
+
+    friend std::ostream& operator<<(std::ostream& O, const sepval_t& sv) {
+        bool first = true;
+        for (const auto& x: sv.v) {
+            if (!first) O << sv.sep;
+            first = false;
+            O << x;
+        }
+        return O;
+    }
+};
+
+template <typename Sep, typename V>
+sepval_t<Sep,V> sepval(const Sep& sep, const V& v) { return sepval_t<Sep,V>(sep, v); }
+
+std::ostream& operator<<(std::ostream& o, const symbol_table& syms) {
+    for (unsigned i = 0; i<syms.size(); ++i) {
+        symbol s = syms[i];
+        if (s.def()) o << s << ": " << s.def() << "\n";
+    }
+    return o;
+}
+
+// Symbolic GE
+
+using symmrow = msparse::mrow<symbol>;
+using symmatrix = msparse::matrix<symbol>;
+
+// Returns q[c]*p - p[c]*q
+template <typename DefineSym>
+symmrow row_reduce(unsigned c, const symmrow& p, const symmrow& q, DefineSym define_sym) {
+    if (p.index(c)==p.npos || q.index(c)==q.npos) throw std::runtime_error("improper row GE");
+
+    symmrow u;
+    symbol x = q[c];
+    symbol y = p[c];
+
+    auto piter = p.begin();
+    auto qiter = q.begin();
+    unsigned pj = piter->first;
+    unsigned qj = qiter->first;
+
+    while (piter!=p.end() || qiter!=q.end()) {
+        unsigned j = std::min(pj, qj);
+        symbol_term t1, t2;
+
+        if (j==pj) {
+            t1 = x*piter->second;
+            ++piter;
+            pj = piter==p.end()? p.npos: piter->first;
+        }
+        if (j==qj) {
+            t2 = y*qiter->second;
+            ++qiter;
+            qj = qiter==q.end()? q.npos: qiter->first;
+        }
+        if (j!=c) {
+            u.push_back({j, define_sym(t1-t2)});
+        }
+    }
+    return u;
+}
+
+// Actual GE reduction.
+// Note: ncol: number of columns before augmentation.
+// Another note: if we're going to preference diagonal pivots
+// (owing to application), then priority queue is overkill.
+template <typename DefineSym>
+void gj_reduce(symmatrix& A, unsigned ncol, DefineSym define_sym) {
+    struct pq_entry {
+        unsigned key; // first non-zero column after pivot column
+        unsigned mincol;
+        unsigned row;
+    };
+
+    struct pq_order_t {
+        bool operator()(const pq_entry& a, const pq_entry& b) const {
+            // The last condition preferences pivots on diagonal elements.
+            return a.key>b.key ||
+                  (a.key==b.key && a.mincol<b.mincol) ||
+                  (a.key==b.key && a.mincol==b.mincol && a.row!=a.key);
+        }
+    };
+
+    std::priority_queue<pq_entry, std::vector<pq_entry>, pq_order_t> pq;
```

Comment:
I like the generality, however we are working with SPD matrices, which don't require pivoting.
I would keep it simple and naive for now (which makes things easier for me to understand ;) ).